main.nf

```diff
@@ -124,76 +132,138 @@ process BWA_ALIGN {
  * Convert the format of the alignment to sorted BAM.
  */
 process SAMTOOLS_SORT {
-  tag{"SAMTOOLS_SORT ${sample_id}"}
-  label 'process_low'
+    tag{"${sample_id}"}
+    label 'process_low'
+    conda 'samtools'
 
-  publishDir("${params.outdir}/bam_align", mode: 'copy')
+    publishDir("${params.outdir}/bam_align", mode: 'copy')
 
-  input:
-  tuple val( sample_id ), path( bam )
+    input:
+    tuple val( sample_id ), path( bam )
 
-  output:
-  tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam" ), emit: sorted_bam
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam" ), emit: sorted_bam
 
-  script:
-  """
-  samtools sort -o "${sample_id}.aligned.sorted.bam" ${bam}
-  """
+    script:
+    """
+    samtools sort -o "${sample_id}.aligned.sorted.bam" ${bam}
+    """
 }
 
 /*
  * Index the BAM file for visualization purpose
  */
 process SAMTOOLS_INDEX {
-    // TODO
+    tag{"${sample_id}"}
+    label 'process_low'
+    conda 'samtools'
+
+    publishDir("${params.outdir}/bam_align", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( bam )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.bai" ), emit: sorted_bam_index
+
+    script:
+    """
+    samtools index ${bam}
+    """
 }
 
 /*
  * Calculate the read coverage of positions in the genome.
- */
+*/
 process BCFTOOLS_MPILEUP {
-    // TODO
+    tag{"${sample_id}"}
+    label 'process_high'
+    conda "bioconda::bcftools=1.17"
+
+    publishDir("${params.outdir}/bcftools_mpileup", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( bam )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.bcf" ), emit: bcf
+
+    script:
+    """
+    bcftools mpileup -O b -o ${sample_id}.aligned.sorted.bam.bcf -f ${params.genome} ${bam}
+    """
 }
 
 /*
  * Detect the single nucleotide variants (SNVs).
- */
+*/
 process BCFTOOLS_CALL {
-    // TODO
+    tag{"${sample_id}"}
+    label 'process_high'
+    conda "bioconda::bcftools=1.17"
+
+    publishDir("${params.outdir}/bcftools_call", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( bcf )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.vcf" ), emit: vcf
+
+    script:
+    """
+    bcftools call -vmO v -o ${sample_id}.aligned.sorted.bam.vcf ${bcf}
+    """
 }
 
-/*
- * Filter and report the SNVs in VCF (variant calling format).
- */
+//* Filter and report the SNVs in VCF (variant calling format).
+
 process VCFUTILS {
-    // TODO
+    tag{"VCFUTILS ${sample_id}"}
+    label 'process_high'
+    conda "bioconda::bcftools=1.17"
+
+    publishDir("${params.outdir}/vcfutils", mode: 'copy')
+
+    input:
+    tuple val( sample_id ), path( vcf )
+
+    output:
+    tuple val( sample_id ), path( "${sample_id}.aligned.sorted.bam.vcf" ), emit: vcf
+
+    script:
+    """
+        vcfutils.pl varFilter -Q 20 -d 10 -D 1000 ${vcf} > ${sample_id}.aligned.sorted.bam.vcf
```

Comment:
The orginal script was 

```sh
    vcfutils.pl varFilter $variants > $final_variants
``` 

This has some extra filtering that's occurring, so your results will vary.